lib/core/src/sync/mod.rs

```diff
@@ -1 +1,189 @@
+        self.fetch_and_save_records().await?;
+
+        // Step 2: Grab all pending incoming records from the database, merge them with
+        // outgoing if necessary, then apply
+        let incoming_records = self.persister.get_incoming_records()?;
+        for new_record in incoming_records {
+            // Step 3: Check whether or not record is updatable (from its schema_version)
+            let is_major_update = unwrap_or_continue!(
+                new_record.is_major_update(),
+                "Could not parse record's `schema_version`: {}"
+            );
+            if is_major_update {
+                continue;
+            }
+
+            // Step 4: Check whether we already have this record, and if the revision is newer
+            let maybe_sync_state = unwrap_or_continue!(
+                self.persister.get_sync_state_by_record_id(&new_record.id),
+                "Could not get sync state: {}"
+            );
+            if let Some(sync_state) = &maybe_sync_state {
+                if sync_state.record_revision > new_record.revision {
+                    continue;
+                }
+            }
+
+            // Step 5: Decrypt the incoming record
+            let mut decrypted_record = unwrap_or_continue!(
+                new_record.decrypt(self.signer.clone()),
+                "Could not decrypt incoming record: {}"
+            );
+
+            // Step 6: Merge with outgoing records, if present
+            let maybe_outgoing_details = unwrap_or_continue!(
+                self.persister
+                    .get_sync_outgoing_details(&decrypted_record.id),
+                "Could not get sync outgoing details: {}"
+            );
+
+            if let Some(outgoing_details) = maybe_outgoing_details {
+                if let Some(updated_fields) = outgoing_details.updated_fields {
+                    let local_data = self.fetch_sync_data(
+                        &decrypted_record.data.id(),
+                        outgoing_details.record_type,
+                    )?;
+                    unwrap_or_continue!(
+                        decrypted_record.data.merge(&local_data, updated_fields),
+                        "Could not merge incoming record with outgoing changes: {}"
+                    );
+                }
+            }
+
+            // Step 7: Apply the changes
+            unwrap_or_continue!(
+                self.apply_record(&decrypted_record),
+                "Could not apply record: {}"
+            );
+
+            // Step 8: Update sync state
+            let update_sync_state_result = self.persister.set_sync_state(SyncState {
+                data_id: decrypted_record.data.id().to_string(),
+                record_id: decrypted_record.id,
+                record_revision: decrypted_record.revision,
+                is_local: maybe_sync_state
+                    .map(|state| state.is_local)
+                    .unwrap_or(false),
+            });
```

Comment:
Updating the final record with the sync state should be in one transaction (for data consistency in case there is a local update right after the read).
Inside the transaction we should do the update only if the sync state wasn't updated since last read.

Reply:
Agreed :+1: